bin/add_attachment.py

```diff
@@ -0,0 +1,103 @@
+#! /usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019 Matt Post <post@cs.jhu.edu>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Used to add attachments to the Anthology.
+Usage:
+
+  add_attachment.py paper_id URL_OR_PATH TYPE
+
+- The ACL ID of the paper (e.g., P17-1012)
+- The path to the attachment (can be a URL)
+- The attachment type (poster, slides, software)
+
+Puts the file in place and modifies the XML.
+"""
+
+import argparse
+import os
+import shutil
+import ssl
+import sys
+import tempfile
+
+from add_revision import maybe_copy
+
+import lxml.etree as ET
+import urllib.request
+
+def main(args):
+
+    if args.path.startswith('http'):
+        _, input_file_path = tempfile.mkstemp()
+        try:
+            print('Downloading file from {}'.format(args.path), file=sys.stderr)
+            with urllib.request.urlopen(args.path) as url, open(input_file_path, mode='wb') as input_file_fh:
+                input_file_fh.write(url.read())
+        except ssl.SSLError:
+            print('An SSL error was encountered in downloading the files.', file=sys.stderr)
+            sys.exit(1)
+    else:
+        input_file_path = args.path
+
+    volume_letter, year, paper_num = [args.paper_id[0], args.paper_id[1:3], args.paper_id[4:]]
+    volume_id = '{}{}'.format(volume_letter, year)
+    ext = args.path.split('.')[-1]
+
+    file_name = f'{args.paper_id}.{args.type.capitalize()}.{ext}'
+    output_dir = os.path.join(args.attachment_root, volume_letter, volume_id)
+
+    # Make sure directory exists
+    if not os.path.exists(output_dir):
+        print('No such directory "{}"'.format(output_dir), file=sys.stderr)
+        sys.exit(1)
+
+    # Copy file
+    dest_path = os.path.join(output_dir, file_name)
+    maybe_copy(input_file_path, dest_path, do=True)
+
+    # Update XML
+    xml_file = os.path.join(os.path.dirname(sys.argv[0]), '..', 'data', 'xml', f'{volume_id}.xml')
+    tree = ET.parse(xml_file)
+    if not tree.getroot().tail: tree.getroot().tail = '\n'
+    for paper in tree.getroot().findall('paper'):
```

Comment:
Minor suggestion: Add an `else` clause to the for-loop that warns/errors when the paper ID isn't found, and maybe move the XML updating before the copying, so in case of a typo the file is not in the wrong place already.

Reply:
I always forget about `for...else`. Great suggestion, thanks.